Provide an initial implementation

rjw57 opened the pull request:
This PR populates this repo with some code. It is best reviewed commit wise.

Some commits merely prepare the ground for the implementation:

* Commit 8eb99b1 strips out web-related functionality from the default boilerplate template and provides a simple "stub" management command, ``gatherstats`` which does nothing other than print ``Hello, world`` to the console.
* Commits e99a75a, c5fa905 and fee8dbe are essentially "plumbing" commits which change some small bit of configuration or install useful packages in the Docker images.

The bulk of the implementation is contained in commits 4da7e61, 04fbf48 and 7148d39:

* Commit 4da7e61 adds a new database model for recording a single statistic.
* Commit 04fbf48 adds functionality to atomically create Statistic models from a single response from the gatherstats endpoint.
* Commit 7148d39 implements the ``gatherstats`` management command which fetches the stats response and uses the functionality in 04fbf48 to update the database.

Travis has been configured at https://travis-ci.org/uisautomation/iar-gatherstats.

Docker hub has been configured with an automatically built image at https://hub.docker.com/r/uisautomation/iar-gatherstats/.

And example of the documentation built from this PR is at https://rjw57.github.io/iar-gatherstats/.

rjw57 committed:
add IAR license file

Add the standard MIT license used for all public IAR code.

rjw57 committed:
replace web-related functionality with stub gatherstats command

This repository will contain only management commands. Remove
configuration required for hosting a web application replace it with a
single management command which can make use of the Django
infrastructure.

The management command is a stub command at the moment in order to
separate functionality and this large removal of code into separate
commits.

rjw57 committed:
add new model: Statistic

Add a new model, Statistic, which stores a single statistic from the
stats endpoint. We explicitly flatten the JSON object returned by the
API and store them as individual rows in the database.

rjw57 committed:
gatherstats.models: add method to create Statistic rows from stats response

Add a helper method to the Statistic manager class which atomically
creates Statistic rows for values from a stats response. Add some basic
tests of functionality and document.

rjw57 committed:
add postgres client to development Docker image

This is very useful if you want to run ./manage_developer.sh dbshell.

rjw57 committed:
implement gatherstats management command

The stub implementation of gatherstats is removed and replaced with an
actual implementation which uses the plumbing created by previous
commits.

The management command takes either a URL to fetch or a local file path.
In both cases the endpoint is set automatically but can be overridden
from the command line if necessary.

rjw57 committed:
provide a default "secret" key

Since this container contains only a single management command, the
secret key is unused. Provide a (non-secret) default and document this
in the README for those to whom this gives the willies.

rjw57 commented:
This PR does not update uisautomation/iar-deploy. That part of the story has moved to uisautomation/iar-deploy#34.

msb commented on gatherstats/models.py:
```diff
@@ -1,3 +1,92 @@
+            # Generate indices for endpoints and fetched at value alone and the
+            # endpoint/key/fetched_at combination since these are likely to be the most common
+            # filtering options. More indexes can be created should it become necessary.
+            models.Index(fields=['endpoint']),
```
as discussed - this can be removed

msb commented on gatherstats/models.py:
```diff
@@ -1,4 +1,73 @@
+    """
+
+    @transaction.atomic
+    def create_from_stats_response(self, endpoint, body, fetched_at=None):
```
could you use ``.., fetched_at=timezone.now()):`` ?

rjw57 replied:
No, because the ``create_from_stats_response()`` almost certainly is not simultaneous with the fetch. By design, I want the caller to be explicit about what time they actually fetched the endpoint at.

rjw57 replied:
... and, additionally, I *think* that would make the default fetched_at time to be the time the module was first imported.

msb commented on README.md:
```diff
@@ -1,9 +1,47 @@
+via GitHub pages.
+
+## Quickstart
+
```
as discussed - add minimal commands for the creation/migration of a DB (either postgres or sqlite - don't mind)

rjw57 committed:
CR: remove redundant index

rjw57 committed:
CR: clarify "Quickstart" section in README

Itwas not clear if the Quickstart section was intended for those using
a deployed gatherstats instance or those wanting to start developing.
Split both use cases into separate sections.
